File: cpp/HLD.cpp
template<typename T, T(*op)(T,T)>
struct HLD {
    private:
    int n;
    vector<vector<int>>& to;
    vector<int> size, par, in, head;
    SegmentTree<T, op> tree;
    T e;

    public:
    HLD(vector<vector<int>>& to, const vector<T>& w, T e)
        :to(to), e(e), tree(to.size(), e), n(to.size()) {
        size.resize(n);
        in.resize(n);
        par.resize(n);
        head.resize(n);

        dfs(0, -1);
        int idx = 0;
        decomp(0, -1, idx);
        for(int i = 0;i < n; ++i){
            tree.set(in[i], w[i]);
        }
        tree.build();
    }

    T get(int v) const {
        assert(0 <= v && v < n);
        return tree.get(in[v]);
    }

    void update(int v, T val) {
        assert(0 <= v && v < n);
        tree.update(in[v], val);
    }

    T query(int u, int v) const {
        T retu = e;
        while(head[u] != head[v]){
            if(in[head[u]] < in[head[v]]){
                swap(u, v);
            }

            retu = op(retu, tree.query(in[head[u]], in[u]+1));
            u = par[head[u]];
        }
        if(in[u] > in[v])swap(u, v);
        retu = op(retu, tree.query(in[u], in[v] + 1));
        return retu;
    }

    private:
    void dfs(int v, int p) {
        par[v] = p;
        size[v] = 1;
        if(to[v].size() && to[v][0] == p){
            swap(to[v][0], to[v].back());
        }
        for(int& nxt: to[v]){
            if(nxt == p)continue;
            dfs(nxt, v);
            size[v] += size[nxt];
            if(size[nxt] > size[to[v][0]]){
                swap(nxt, to[v][0]);
            }
        }
    }

    void decomp(int v, int par, int &idx) {
        in[v] = idx;
        ++idx;
        for(int nxt: to[v]){
            if(par == nxt)continue;
            if(to[v][0] == nxt){
                head[nxt] = head[v];
            }else{
                head[nxt] = nxt;
            }
            decomp(nxt, v, idx);
        }
    }
};

File: rust/HLD.rs
struct HLD<M: Monoid> {
    to: Vec<Vec<usize>>,
    head: Vec<usize>,
    par: Vec<usize>,
    to_seg_idx: Vec<usize>,
    tree: SegmentTree<M>,
}

#[allow(dead_code)]
impl<M: Monoid> HLD<M> {
    fn new(to: Vec<Vec<usize>>, w: &Vec<M>) -> Self {
        let n = to.len();
        let mut hld = Self {
            to,
            head: vec![0; n],
            par: vec![0; n],
            to_seg_idx: vec![0; n],
            tree: SegmentTree::new(n),
        };

        let mut size = vec![0; n];
        hld.dfs(0, n, &mut size);
        let mut idx = 0;
        hld.decomp(0, n, &mut idx);

        for i in 0..n {
            hld.tree.set(hld.to_seg_idx[i], w[i].clone());
        }
        hld.tree.build();
        hld
    }

    fn get(&self, v: usize) -> M {
        self.tree.get(self.to_seg_idx[v])
    }

    fn query(&mut self, mut u: usize, mut v: usize) -> M {
        let mut retu = M::e();
        while self.head[u] != self.head[v] {
            if self.to_seg_idx[self.head[u]] < self.to_seg_idx[self.head[v]] {
                std::mem::swap(&mut u, &mut v);
            }

            retu = M::op(&retu, &self.tree.query(self.to_seg_idx[self.head[u]], self.to_seg_idx[u] + 1));
            u = self.par[self.head[u]];
        }

        let l =  self.to_seg_idx[u].min(self.to_seg_idx[v]);
        let r = self.to_seg_idx[u].max(self.to_seg_idx[v]);
        M::op(&retu, &self.tree.query(l, r +1))
    }

    fn update(&mut self, i: usize, w: M) {
        self.tree.update(self.to_seg_idx[i], w);
    }

    fn dfs(&mut self, v: usize, par: usize, size: &mut Vec<usize>) {
        let n = self.to[v].len();
        if self.to[v].len() >= 2 && self.to[v][0] == par {
            self.to[v].swap(0, 1);
        }
        size[v] = 1;
        for i in 0..n {
            let nxt = self.to[v][i];
            if nxt == par { continue; }
            self.par[nxt] = v;
            self.dfs(nxt, v, size);
            size[v] += size[nxt];

            if size[nxt] > size[self.to[v][0]] {
                self.to[v].swap(i, 0);
            }
        }
    }

    fn decomp(&mut self, v: usize, par: usize, idx: &mut usize) {
        self.to_seg_idx[v] = *idx;
        *idx += 1;
        for i in 0..self.to[v].len() {
            let nxt = self.to[v][i];
            if nxt == par { continue; }
            if i == 0 {
                self.head[nxt] = self.head[v];
            } else {
                self.head[nxt] = nxt;
            }
            self.decomp(nxt, v, idx);
        }
    }
}
